lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -1308,87 +1368,41 @@ export class UserAgentApplication {
     if (isWindowOpenerMsal) {
       self = window.opener.msal;
       isPopup = true;
-    }
-    else if (window.parent && window.parent.msal) {
+    } else if (window.parent && window.parent.msal) {
       self = window.parent.msal;
     }
 
-    // if (window.parent !== window), by using self, window.parent becomes equal to window in getRequestInfo method specifically
-    // TODO: Refactor so that this or saveTokenFromHash function returns a response object
-    const requestInfo = self.getRequestInfo(hash);
+    // if (window.parent !== window), by using self, window.parent becomes equal to window in getResponseState method specifically
+    const stateInfo = self.getResponseState(hash);
 
-    let token: string = null;
-    let tokenReceivedCallback: (errorDesc: string, token: string, error: string, tokenType: string) => void = null;
-    let tokenType: string;
+    let tokenResponseCallback: (response: AuthResponse, error: AuthError) => void = null;
 
     self.logger.info("Returned from redirect url");
-    // If parent window is the msal instance which opened the current window
-    if (window.parent !== window && window.parent.msal) {
-        tokenReceivedCallback = window.parent.callBackMappedToRenewStates[requestInfo.stateResponse];
+    // If parent window is the msal instance which opened the current window (iframe)
+    if (this.parentIsMsal()) {
+        tokenResponseCallback = window.parent.primaryMsalCallbackMappedToRenewStates[stateInfo.state];
     }
-    // Current window is window opener
+    // Current window is window opener (popup)
     else if (isWindowOpenerMsal) {
-        tokenReceivedCallback = window.opener.callBackMappedToRenewStates[requestInfo.stateResponse];
+        tokenResponseCallback = window.opener.callBackMappedToRenewStates[stateInfo.state];
     }
-    // Other cases
+    // Redirect cases
     else {
+      tokenResponseCallback = null;
       // if set to navigate to loginRequest page post login
-      if (self._navigateToLoginRequestUrl) {
-        tokenReceivedCallback = null;
-        self._cacheStorage.setItem(Constants.urlHash, hash);
+      if (self.navigateToLoginRequestUrl) {
+        self.cacheStorage.setItem(Constants.urlHash, hash);
         if (window.parent === window && !isPopup) {
-          window.location.href = self._cacheStorage.getItem(Constants.loginRequest, this.inCookie);
+          window.location.href = self.cacheStorage.getItem(Constants.loginRequest, this.inCookie);
         }
         return;
       }
-      // Close current window??
       else {
-        tokenReceivedCallback = self._tokenReceivedCallback;
```

Comment:
This is not needed I assume because we are passing the callback in processCallBack()?

Reply:
correct